autoscale/src/main/java/com/sequenceiq/periscope/service/YarnInvocationService.java

```diff
@@ -0,0 +1,138 @@
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.StackV4Response;
+import com.sequenceiq.periscope.domain.Cluster;
+import com.sequenceiq.periscope.domain.LoadAlert;
+import com.sequenceiq.periscope.domain.LoadAlertConfiguration;
+import com.sequenceiq.periscope.model.ScalingAdjustmentType;
+import com.sequenceiq.periscope.model.yarn.YarnScalingServiceV1Response;
+import com.sequenceiq.periscope.monitor.client.YarnMetricsClient;
+import com.sequenceiq.periscope.monitor.evaluator.load.YarnResponseUtils;
+import com.sequenceiq.periscope.monitor.sender.ScalingEventSender;
+import com.sequenceiq.periscope.utils.ClusterUtils;
+import com.sequenceiq.periscope.utils.StackResponseUtils;
+import com.sequenceiq.periscope.utils.TimeUtil;
+
+@Service
+public class YarnInvocationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(YarnInvocationService.class);
+
+    @Inject
+    private YarnResponseUtils yarnResponseUtils;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private ScalingEventSender eventSender;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    private LoadAlertConfiguration loadAlertConfiguration;
+
+    private String policyHostGroup;
+
+    public void pollYarnMetricsAndScaleCluster(Long clusterId, String pollingUserCrn, boolean stopStartEnabled, StackV4Response stackV4Response)
+            throws Exception {
+
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+
+        Map<String, String> hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup);
+        List<String> stoppedHostInstanceIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackV4Response, policyHostGroup);
+
+        int serviceHealthyHostGroupSize = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackV4Response, policyHostGroup).size();
+        int existingHostGroupSize = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup).size();
+
+        int maxResourceValueOffset = loadAlertConfiguration.getMaxResourceValue() - (stopStartEnabled ? serviceHealthyHostGroupSize : existingHostGroupSize);
+        int minResourceValueOffset = serviceHealthyHostGroupSize - loadAlertConfiguration.getMinResourceValue();
+
+        int maxAllowedUpScale = Math.max(maxResourceValueOffset, 0);
+        int maxAllowedDownScale = Math.max(minResourceValueOffset, 0);
```

Comment:
A lot of this code above seems unnecessary. By this point - we know that we're scaling purely based on YARN recommendation, and ensuring we don't go past limits. There should be nothing beyond that.

Reply:
These params are used to calculate the limits for upscale and downscale, along with calculating the downscaleFactor parameter which is sent to YARN as part of the API request. I have simplified the code below to determine these limits outside of YarnResponseUtils